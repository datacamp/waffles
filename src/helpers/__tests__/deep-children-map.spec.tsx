import React, { isValidElement, cloneElement } from 'react';
import { render } from '@testing-library/react';

import { deepChildrenMap } from '../index';

type DeepCloneProps = {
  children: React.ReactNode;
};

function DeepClone({ children }: DeepCloneProps) {
  return (
    <div>
      {deepChildrenMap(children, (child) => {
        if (isValidElement(child)) {
          return cloneElement(child, {
            'data-testid': 'taylor-swift-clone',
          });
        }

        return null;
      })}
    </div>
  );
}

describe('deepChildrenMap', () => {
  it('modifies each nested element', () => {
    const { getAllByTestId } = render(
      <DeepClone>
        <div>
          <h1>Little Test</h1>
          <div>
            <span>Test</span>
          </div>
          <div>
            <span>Test2</span>
          </div>
          <p>Test 3</p>
        </div>
      </DeepClone>,
    );

    const clonedElements = getAllByTestId('taylor-swift-clone');

    expect(clonedElements).toHaveLength(7);
  });

  it('renders snapshot', () => {
    const { container } = render(
      <DeepClone>
        <div>
          <h1>Little Test</h1>
          <div>
            <span>Test</span>
          </div>
          <p>Test 3</p>
        </div>
      </DeepClone>,
    );

    const elements = container.firstChild;

    expect(elements).toMatchSnapshot();
  });
});
