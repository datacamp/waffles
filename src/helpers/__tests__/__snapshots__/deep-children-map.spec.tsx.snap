// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deepChildrenMap renders snapshot 1`] = `
<div>
  <div
    data-testid="taylor-swift-clone"
  >
    <h1
      data-testid="taylor-swift-clone"
    >
      Little Test
    </h1>
    <div
      data-testid="taylor-swift-clone"
    >
      <span
        data-testid="taylor-swift-clone"
      >
        Test
      </span>
    </div>
    <p
      data-testid="taylor-swift-clone"
    >
      Test 3
    </p>
  </div>
</div>
`;
